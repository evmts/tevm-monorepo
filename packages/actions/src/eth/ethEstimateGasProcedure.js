import { callProcedure } from '../call/callProcedure.js'

/**
 * Request handler for eth_estimateGas JSON-RPC requests.
 * @param {import('@tevm/node').TevmNode} client
 * @returns {import('./EthProcedure.js').EthEstimateGasJsonRpcProcedure}
 */
export const ethEstimateGasJsonRpcProcedure = (client) => {
	return async (request) => {
		const estimateGasRequest = /** @type {import('./EthJsonRpcRequest.js').EthEstimateGasJsonRpcRequest}*/ (request)
		const [_params, blockTag, stateOverrides, blockOverrides] = estimateGasRequest.params

		const getParams = () => {
			/**
			 * @type {import('../call/CallJsonRpcRequest.js').CallJsonRpcRequest['params']}
			 */
			const params = [
				{
					..._params,
					...(blockTag !== undefined ? { blockTag } : {}),
				},
			]
			if (blockOverrides !== undefined) {
				params.push(stateOverrides ?? {}, blockOverrides)
			}
			if (stateOverrides !== undefined) {
				params.push(...params, stateOverrides)
			}
			return params
		}

		const callResult = await callProcedure(client)({
			...estimateGasRequest,
			params: getParams(),
			method: 'tevm_call',
		})
		if (callResult.error || !callResult.result) {
			return {
				...callResult,
				method: estimateGasRequest.method,
			}
		}
		return {
			method: estimateGasRequest.method,
			result: callResult.result.totalGasSpent ?? callResult.result.executionGasUsed,
			jsonrpc: '2.0',
			...(estimateGasRequest.id ? { id: estimateGasRequest.id } : {}),
		}
	}
}
