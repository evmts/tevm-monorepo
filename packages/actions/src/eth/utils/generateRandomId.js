/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
