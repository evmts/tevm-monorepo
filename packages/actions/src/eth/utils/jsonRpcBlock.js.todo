import { bytesToHex } from "viem"

const ethjsTxToTevmTx = (tx) => {

}

/**
 * Transoforms an ethjs block to a tevm Block
 * @param {import("@ethereumjs/block").Block} block
 * @param {boolean} includeTransactions
 * @returns {Promise<import("@tevm/actions-types").Block>}
 */
const ethjsToTevmBlock = async (
  block,
  includeTransactions,
) => {
  const transactions = block.transactions.map((tx, txIndex) =>
    includeTransactions ? ethjsTxToTevmTx(tx, block, txIndex) : bytesToHex(tx.hash())
  )
  const withdrawalsAttr =
    block.header.withdrawalsRoot !== undefined
      ? {
        withdrawalsRoot: bytesToHex(block.header.withdrawalsRoot),
        withdrawals: block.toJSON().withdrawals,
      }
      : {}
  const td = await chain.getTd(block.hash(), block.header.number)
  return {
    coinbase: block.toJSON().header?.coinbase
    number: block.toJSON().header.number,
    hash: bytesToHex(block.hash()),
    parentHash: block.toJSON().header.parentHash,
    mixHash: block.toJSON().header.mixHash,
    nonce: block.toJSON().header.nonce,
    sha3Uncles: block.toJSON().header.uncleHash,
    logsBloom: block.toJSON().header.logsBloom,
    transactionsRoot: block.toJSON().header.transactionsTrie,
    stateRoot: block.toJSON().header.stateRoot,
    receiptsRoot: block.toJSON().header.receiptTrie,
    miner: block.toJSON().header.coinbase,
    difficulty: block.toJSON().header.difficulty,
    totalDifficulty: bigIntToHex(td),
    extraData: block.toJSON().header.extraData,
    size: intToHex(utf8ToBytes(JSON.stringify(block.toJSON())).byteLength),
    gasLimit: block.toJSON().header.gasLimit,
    gasUsed: block.toJSON().header.gasUsed,
    timestamp: block.toJSON().header.timestamp,
    transactions,
    uncles: block.uncleHeaders.map((uh) => bytesToHex(uh.hash())),
    baseFeePerGas: block.toJSON().header.baseFeePerGas,
    ...withdrawalsAttr,
    blobGasUsed: block.toJSON().header.blobGasUsed,
    excessBlobGas: block.toJSON().header.excessBlobGas,
    parentBeaconBlockRoot: block.toJSON().header.parentBeaconBlockRoot,
  }
}
