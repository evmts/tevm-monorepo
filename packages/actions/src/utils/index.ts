export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
