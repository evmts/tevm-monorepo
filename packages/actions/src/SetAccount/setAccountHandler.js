import { createAddress } from '@tevm/address'
import { AccountNotFoundError, InternalError } from '@tevm/errors'
import { EthjsAccount } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateSetAccountParams } from './validateSetAccountParams.js'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./SetAccountHandlerType.js').SetAccountHandler}
 */
export const setAccountHandler =
	(client, options = {}) =>
	async (params) => {
		const { throwOnFail = options.throwOnFail ?? true } = params
		/**
		 * @type {Array<import('../SetAccount/TevmSetAccountError.js').TevmSetAccountError>}
		 */
		const errors = validateSetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}

		const address = createAddress(params.address)

		/**
		 * @type {Array<Promise<any>>}
		 */
		const promises = []
		try {
			const vm = await client.getVm()

			// check if account exists
			const account = await getAccountHandler(client)({ ...params, throwOnFail: false })
			if (account.errors?.length && !(account.errors[0] instanceof AccountNotFoundError)) {
				client.logger.error('there was an unexpected error getting account', account.errors)
				throw account.errors.length > 1 ? new AggregateError(account.errors) : account.errors[0]
			}

			promises.push(
				vm.stateManager.putAccount(
					address,
					new EthjsAccount(
						params.nonce ?? account?.nonce,
						params.balance ?? account?.nonce,
						(params.storageRoot && hexToBytes(params.storageRoot)) ??
							(account?.storageRoot !== undefined && account?.storageRoot !== '0x'
								? hexToBytes(account.storageRoot)
								: undefined),
						(params.deployedBytecode && hexToBytes(keccak256(params.deployedBytecode))) ??
							(account?.deployedBytecode !== undefined ? hexToBytes(keccak256(account.deployedBytecode)) : undefined),
					),
				),
			)
			if (params.deployedBytecode) {
				promises.push(vm.stateManager.putContractCode(address, hexToBytes(params.deployedBytecode)))
			}
			// state clears state first stateDiff does not
			if (params.state) {
				await vm.stateManager.clearContractStorage(address)
			}
			const state = params.state ?? params.stateDiff
			if (state) {
				for (const [key, value] of Object.entries(state)) {
					promises.push(
						vm.stateManager.putContractStorage(
							address,
							hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (key), { size: 32 }),
							hexToBytes(value),
						),
					)
				}
			}
			const results = await Promise.allSettled(promises)
			for (const result of results) {
				if (result.status === 'rejected') {
					errors.push(new InternalError('Unable to put storage', { cause: result.reason }))
				}
			}

			if (errors.length > 0) {
				return maybeThrowOnFail(throwOnFail, { errors })
			}
			await vm.stateManager.checkpoint()
			await vm.stateManager.commit(false)

			return {}
		} catch (e) {
			errors.push(new InternalError('Unexpected error setting account', { cause: e }))
			return maybeThrowOnFail(throwOnFail, { errors })
		}
	}
