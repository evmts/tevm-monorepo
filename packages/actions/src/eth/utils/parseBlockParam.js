import { InvalidBlockError } from '@tevm/errors'
import { hexToBigInt, hexToBytes } from '@tevm/utils'

/**
 * @param {import('@tevm/blockchain').Chain} blockchain
 * @param {import('../../common/BlockParam.js').BlockParam} blockParam
 * @returns {Promise<bigint >}
 */
export const parseBlockParam = async (blockchain, blockParam) => {
	if (typeof blockParam === 'number') {
		return BigInt(blockParam)
	}
	if (typeof blockParam === 'bigint') {
		return blockParam
	}
	if (typeof blockParam === 'string' && blockParam.startsWith('0x')) {
		if (blockParam.length === 66) {
			const block = await blockchain.getBlock(hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (blockParam)))
			return BigInt(block.header.number)
		}
		return hexToBigInt(/** @type {import('@tevm/utils').Hex}*/ (blockParam))
	}
	if (blockParam === 'safe') {
		const safeBlock = blockchain.blocksByTag.get('safe')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('safe not currently supported as block tag')
	}
	if (blockParam === 'latest' || blockParam === undefined) {
		const safeBlock = blockchain.blocksByTag.get('latest')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('latest block does not exist on chain')
	}
	if (blockParam === 'pending') {
		// for pending we need to mine a new block and then handle it
		// let's skip this functionality for now
		throw new InvalidBlockError(
			'Pending not yet supported but will be in future. Consider opening an issue or reaching out on telegram if you need this feature to expediate its release',
		)
	}
	if (blockParam === 'earliest') {
		return BigInt(1)
	}
	if (blockParam === 'finalized') {
		throw new InvalidBlockError('finalized noet yet supported for this feature')
	}
	blockchain.logger.error({ blockParam }, 'Unknown block param pased to blockNumberHandler')
	throw new InvalidBlockError(`Unknown block param ${blockParam} pased to blockNumberHandler`)
}
