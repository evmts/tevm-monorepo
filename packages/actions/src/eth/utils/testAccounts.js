import { mnemonicToAccount } from '@tevm/utils'

const oneToTen = Array.from(Array(10).keys())

export const mnemonic =
	'test test test test test test test test test test test junk'

export const testAccounts =
	/**
	 * @type {[import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount]}
	 */
	(oneToTen.map((i) => mnemonicToAccount(mnemonic, { addressIndex: i })))
