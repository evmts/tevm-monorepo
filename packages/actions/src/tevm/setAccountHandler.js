import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAccount, EthjsAddress } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { validateSetAccountParams } from '@tevm/zod'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'vm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').SetAccountHandler}
 */
export const setAccountHandler = (client, options = {}) => async (params) => {
	const { throwOnFail = options.throwOnFail ?? true } = params
	/**
	 * @type {Array<import('@tevm/errors').SetAccountError>}
	 */
	const errors = validateSetAccountParams(params)
	if (errors.length > 0) {
		return maybeThrowOnFail(throwOnFail, { errors })
	}

	const address = new EthjsAddress(hexToBytes(params.address))
	try {
		await client.vm.stateManager.putAccount(
			address,
			new EthjsAccount(
				params.nonce,
				params.balance,
				params.storageRoot && hexToBytes(params.storageRoot),
				params.deployedBytecode &&
					hexToBytes(keccak256(params.deployedBytecode)),
			),
		)
		if (params.deployedBytecode) {
			await client.vm.stateManager.putContractCode(
				address,
				hexToBytes(params.deployedBytecode),
			)
		}
		await client.vm.stateManager.checkpoint()
		await client.vm.stateManager.commit()
		// TODO offer way of setting contract storage with evm.stateManager.putContractStorage
		return {}
	} catch (e) {
		errors.push(
			createError(
				'UnexpectedError',
				typeof e === 'string'
					? e
					: e instanceof Error
					? e.message
					: 'unknown error',
			),
		)
		return maybeThrowOnFail(throwOnFail, { errors })
	}
}
