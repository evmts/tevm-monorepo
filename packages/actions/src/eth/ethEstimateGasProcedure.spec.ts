import { createAddress } from '@tevm/address'
import { type TevmNode, createTevmNode } from '@tevm/node'
import { ErrorContract } from '@tevm/test-utils'
import { PREFUNDED_ACCOUNTS, encodeFunctionData, numberToHex } from '@tevm/utils'
import { beforeEach, describe, expect, it } from 'vitest'
import { setAccountHandler } from '../SetAccount/setAccountHandler.js'
import type { EthEstimateGasJsonRpcRequest } from './EthJsonRpcRequest.js'
import { ethEstimateGasJsonRpcProcedure } from './ethEstimateGasProcedure.js'

let client: TevmNode

beforeEach(() => {
	client = createTevmNode()
})

describe('ethEstimateGasJsonRpcProcedure', () => {
	it('should estimate gas successfully', async () => {
		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 1,
			params: [
				{
					from: '0x0000000000000000000000000000000000000000',
					to: '0x0000000000000000000000000000000000000000',
					data: '0x',
				},
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_estimateGas')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})
	it('should handle block tag', async () => {
		const latestBlock = await client.getVm().then((vm) => vm.blockchain.getCanonicalHeadBlock())
		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 1,
			params: [
				{
					from: '0x0000000000000000000000000000000000000000',
					to: '0x0000000000000000000000000000000000000000',
					data: '0x',
				},
				numberToHex(latestBlock.header.number),
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_estimateGas')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle errors from callProcedure', async () => {
		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 1,
			params: [
				{
					to: '0x0000000000000000000000000000000000000000',
					from: '0x0000000000000000000000000000000000000000',
					data: '0xINVALID_DATA',
				},
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)
		expect(response.error).toBeDefined()
		expect(response.result).toBeUndefined()
		expect(response.method).toBe('eth_estimateGas')
		expect(response.id).toBe(request.id as any)
		expect(response.error).toMatchSnapshot()
	})

	it('should handle requests without an id', async () => {
		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			params: [
				{
					to: '0x0000000000000000000000000000000000000000',
					from: '0x0000000000000000000000000000000000000000',
					data: '0x',
				},
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_estimateGas')
		expect(response.id).toBeUndefined()
		expect(response.result).toMatchSnapshot()
	})

	it('should handle error responses when using stateOverrides', async () => {
		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 2,
			params: [
				{
					to: '0x0000000000000000000000000000000000000000',
					from: '0x0000000000000000000000000000000000000000',
					data: '0x',
				},
				'latest',
				{
					'0x0000000000000000000000000000000000000000': {
						balance: '0x1000000000000000000',
					},
				},
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)
		expect(response.error).toBeDefined()
		expect(response.result).toBeUndefined()
		expect(response.method).toBe('eth_estimateGas')
		expect(response.id).toBe(2)
		expect(response.error?.code).toBeDefined()
		expect(response.error?.message).toBeDefined()
	})

	it('should handle error responses when using stateOverrides and blockOverrides', async () => {
		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 3,
			params: [
				{
					to: '0x0000000000000000000000000000000000000000',
					from: '0x0000000000000000000000000000000000000000',
					data: '0x',
				},
				'latest',
				{
					'0x0000000000000000000000000000000000000000': {
						balance: '0x1000000000000000000',
					},
				},
				{
					baseFee: '0x1000',
				},
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)
		expect(response.error).toBeDefined()
		expect(response.result).toBeUndefined()
		expect(response.method).toBe('eth_estimateGas')
		expect(response.id).toBe(3)
		expect(response.error?.code).toBeDefined()
		expect(response.error?.message).toBeDefined()
	})

	it('should handle basic error decoding when a contract call reverts', async () => {
		const contractAddress = createAddress('0x1234')
		await setAccountHandler(client)({
			address: contractAddress.toString(),
			deployedBytecode: ErrorContract.deployedBytecode,
		})

		const data = encodeFunctionData({
			abi: ErrorContract.abi,
			functionName: 'revertWithRequireNoMessage',
		})

		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 4,
			params: [
				{
					to: contractAddress.toString(),
					from: '0x0000000000000000000000000000000000000000',
					data,
				},
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)
		expect(response.error).toMatchObject({
			code: 3,
			message: 'execution reverted',
			data: {
				data: '0x',
				errors: [
					'revert\n' +
						'\n' +
						'Docs: https://tevm.sh/reference/tevm/errors/classes/reverterror/\n' +
						'Details: {"error":"revert","errorType":"EVMError"}\n' +
						'Version: 1.1.0.next-73',
				],
			},
		})
		expect(response.result).toBeUndefined()
		expect(response.method).toBe('eth_estimateGas')
		expect(response.id).toBe(4)
	})

	it('should not create transactions or trigger mining during gas estimation', async () => {
		const initialBlock = await client.getVm().then((vm) => vm.blockchain.getCanonicalHeadBlock())
		const initialBlockNumber = initialBlock.header.number

		const txPool = await client.getTxPool()
		const initialTxPoolSize = txPool.pool.size

		const request: EthEstimateGasJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 1,
			params: [
				{
					from: PREFUNDED_ACCOUNTS[0].address,
					to: '0x0000000000000000000000000000000000000000',
					data: '0x',
					nonce: '0x1',
				},
			],
		}

		const response = await ethEstimateGasJsonRpcProcedure(client)(request)

		const finalBlock = await client.getVm().then((vm) => vm.blockchain.getCanonicalHeadBlock())
		const finalBlockNumber = finalBlock.header.number
		const finalTxPoolSize = txPool.pool.size

		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(finalBlockNumber).toBe(initialBlockNumber)
		expect(finalTxPoolSize).toBe(initialTxPoolSize)
	})
})
