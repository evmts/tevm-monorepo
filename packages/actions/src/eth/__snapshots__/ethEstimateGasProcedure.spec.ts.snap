// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ethEstimateGasJsonRpcProcedure > should estimate gas successfully 1`] = `"0x5208"`;

exports[`ethEstimateGasJsonRpcProcedure > should handle block tag 1`] = `"0x5208"`;

exports[`ethEstimateGasJsonRpcProcedure > should handle errors from callProcedure 1`] = `
{
  "code": -32602,
  "data": {
    "errors": [
      "value must be a hex string

Docs: https://tevm.sh/reference/tevm/errors/classes/invaliddataerror/
Version: 1.1.0.next-73",
    ],
  },
  "message": "value must be a hex string

Docs: https://tevm.sh/reference/tevm/errors/classes/invaliddataerror/
Version: 1.1.0.next-73",
}
`;

exports[`ethEstimateGasJsonRpcProcedure > should handle requests without an id 1`] = `"0x5208"`;
